### PR TITLE
Pass in optional arguments by ref

### LINKED ISSUE #23691: Optional `in` parameters are passed by value
Originally filed by @ufcpp on #23636

```csharp
using System;

class Program
{
    static void Main()
    {
        // buffer overrun
        A(1);

        // throw NullReference or AccessViolation Exception
        B(1, 2);
    }

    static void A(in double x = 1, in string y = "") => Console.WriteLine(y);
    static double B(in float x, in float y, in float z = 1.0f) => x * y * z;

}
```

### PR DESCRIPTION
**Pull request by OmarTawfik**
### Customer scenario

Not passing values to optional parameters of `RefKind.In` will emit invalid IL, causing runtime exceptions.

Code this fixes:

```C#
using System;

class Program
{
    static void Main()
    {
        // buffer overrun
        A(1);

        // throw NullReference or AccessViolation Exception
        B(1, 2);
    }

    static void A(in double x = 1, in string y = "") => Console.WriteLine(y);
    static double B(in float x, in float y, in float z = 1.0f) => x * y * z;

}
```

### Bugs this fixes

Fixes #23691

### Workarounds, if any

None.

### Risk

Low. This is fixing an already failing scenario.

### Performance impact

None.

### Is this a regression from a previous update?

This fixes a bug introduced by an earlier PR in the same release, and adds appropriate tests.

### Root cause analysis

An earlier change was made to distinguish between arguments passed with no modifier to an `in` parameter, and arguments passed explicitly by `in` at the call site. This change refactored the local rewriter, which introduced the bug. The compiler (now in master) ignores the refkind of optional parameters when trying to generate the values for non-existing arguments, and passes them by value.

### How was the bug found?

Customer reported on github.

### Test documentation updated?

--

### CODE REVIEW

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler for review.

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler for review

**Comment by VSadov** [comment]
Looks like the bug was the result of adding support for “in” at callsute to mean “strict”.
 Now we consider argument ref kinds when present when emitting calls and omitted optional arguments were not adding their actual ref kinds.

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler can I get a second review?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs [hunk 1]
```diff
@@ -2066,5 +2066,370 @@ static void Main(string[] args)
...
+class Program
+{
+    static void Test(in int value = 5)
+    {
+        System.Console.WriteLine(value);
+    }
+
+    static void Main(string[] args)
+    {
+        Test();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs]: >Test() [](start = 8, length = 6)

Please add verification of IOperation tree for both invocations.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs [hunk 2]
```diff
@@ -2066,5 +2066,370 @@ static void Main(string[] args)
...
+            System.Console.WriteLine($""set p1={p1} p2={p2} to {value}"");
+        }
+    }
+
+    static void Main(string[] args)
+    {
+        var obj = new Program();
+
+        obj[3] += 10;
+        obj[4, 5] += 11;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs]: >obj[4, 5] += 11; [](start = 8, length = 16)

Please add verification of IOperation tree for both indexing operations.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs [hunk 3]
```diff
@@ -2066,5 +2067,772 @@ static void Main(string[] args)
             CompileAndVerify(code, additionalRefs: new[] { reference.ToMetadataReference() }, expectedOutput: "3");
             CompileAndVerify(code, additionalRefs: new[] { reference.EmitToImageReference() }, expectedOutput: "3");
         }
+
+        [Fact]
+        [WorkItem(23691, "https://github.com/dotnet/roslyn/issues/23691")]
+        public void OptionalInParameters_Method_Optional_NoArgs()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs]: >OptionalInParameters_Method_Optional_NoArgs [](start = 20, length = 43)

Consider adding ```[CompilerTrait(CompilerFeature.IOperation)]``` to this and other tests verifying IOperation

**Comment by OmarTawfik** [comment]
@jaredpar for ask mode approval.

**Comment by jaredpar** [comment]
Approved

**Pull request merged**